src/dataflow/src/sink/kafka.rs

```diff
@@ -185,13 +185,58 @@ impl SinkMetrics {
     }
 }
 
+#[derive(Clone)]
+pub struct KafkaSinkSendRetryManager {
+    // Because we flush this fully before moving onto the next timestamp, this queue cannot contain
+    // more messages than are contained in a single timely timestamp.  It's also unlikely to get
+    // beyond the size of the rdkafka internal message queue because that seems to stop accepting
+    // new messages when erorrs are being returned by the delivery callback -- though we should not
+    // rely on that fact if/when we eventually (tm) get around to making our queues bounded.
+    //
+    // If perf becomes an issue, we can do something slightly more complex with a crossbeam channel
+    q: VecDeque<OwnedMessage>,
+    outstanding_send_count: u64,
+}
+
+impl KafkaSinkSendRetryManager {
+    fn new() -> Self {
+        Self {
+            q: VecDeque::new(),
+            outstanding_send_count: 0,
+        }
+    }
+    fn record_send(&mut self) {
+        self.outstanding_send_count += 1;
+    }
+    fn record_error(&mut self, msg: OwnedMessage) {
+        self.q.push_back(msg);
+        self.outstanding_send_count -= 1;
```

Comment:
we could probably mark this struct as `!Sync`, as logically it cannot be used with a lock, but its not-`pub` so should be fine